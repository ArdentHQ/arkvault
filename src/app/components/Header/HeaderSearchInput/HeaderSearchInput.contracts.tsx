export interface HeaderSearchInputProperties {
	placeholder?: string;
	onSearch?: (query: string) => void;
	onReset?: () => void;
	maxLength?: number;
	debounceTimeout?: number;
	defaultQuery?: string;
	resetFields?: boolean;
}
