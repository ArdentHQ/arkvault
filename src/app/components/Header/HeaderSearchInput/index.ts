export * from "./HeaderSearchInput";
export * from "./HeaderSearchInput.contracts";
