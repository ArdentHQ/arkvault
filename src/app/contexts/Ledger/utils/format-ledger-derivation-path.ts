import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
