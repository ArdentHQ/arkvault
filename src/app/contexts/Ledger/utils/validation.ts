import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
