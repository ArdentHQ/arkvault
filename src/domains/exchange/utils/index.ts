export * from "./assertions";
export * from "./errors";
