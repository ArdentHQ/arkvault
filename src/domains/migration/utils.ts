import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
