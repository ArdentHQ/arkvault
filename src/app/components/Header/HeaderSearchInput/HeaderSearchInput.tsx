import cn from "classnames";
import React, { ChangeEvent, FC, useCallback, useEffect, useRef, useState } from "react";
import { useTranslation } from "react-i18next";
import tw, { styled } from "twin.macro";

import { HeaderSearchInputProperties } from "./HeaderSearchInput.contracts";
import { Icon } from "@/app/components/Icon";
import { Input } from "@/app/components/Input";
import { useDebounce } from "@/app/hooks";

const SearchBarInputWrapper = styled.div`
	${tw`xs:[min-width:300px] sm:[min-width:448px] dark:border dark:border-theme-secondary-800`}
`;

export const HeaderSearchInput: FC<HeaderSearchInputProperties> = ({
	placeholder,
	onSearch,
	onReset,
	defaultQuery = "",
	debounceTimeout = 500,
	resetFields = false,
	maxLength,
}) => {
	const { t } = useTranslation();

	const [query, setQuery] = useState(defaultQuery);

	const reference = useRef(null);

	const [debouncedQuery] = useDebounce(query, debounceTimeout);
	useEffect(() => onSearch?.(debouncedQuery), [debouncedQuery]); // eslint-disable-line react-hooks/exhaustive-deps

	const handleQueryReset = useCallback(() => {
		setQuery("");
		onReset?.();
	}, [onReset]);

	useEffect(() => {
		if (resetFields) {
			handleQueryReset();
		}
	}, [resetFields, handleQueryReset]);

	return (
		<div data-testid="HeaderSearchInput" className="relative">
			<SearchBarInputWrapper
				data-testid="HeaderSearchInput__input"
				ref={reference}
				className="flex items-center overflow-hidden rounded-lg border border-theme-secondary-400 bg-theme-background px-4 text-base dark:border-theme-secondary-700"
			>
				<button
					data-testid="HeaderSearchInput__input__reset"
					className={cn("transition-all duration-300 focus:outline-none", { "mr-4": query !== "" })}
					onClick={handleQueryReset}
					type="button"
				>
					<Icon
						className={cn(
							"text-theme-text transition-all duration-300",
							{ "w-0": query === "" },
							{ "w-4": query !== "" },
						)}
						name="Cross"
						size="md"
					/>
				</button>

				<div className="flex-1">
					<Input
						data-testid="HeaderSearchInput__input__input"
						className="-ml-4"
						placeholder={placeholder || `${t("COMMON.SEARCH")}...`}
						value={query}
						maxLength={maxLength}
						isFocused
						ignoreContext
						onChange={(event: ChangeEvent<HTMLInputElement>) => setQuery(event.target.value)}
						noBorder
						noShadow
					/>
				</div>

				<Icon
					className="text-theme-secondary-700 dark:text-theme-secondary-600"
					name="MagnifyingGlassAlt"
					size="lg"
				/>
			</SearchBarInputWrapper>
		</div>
	);
};
