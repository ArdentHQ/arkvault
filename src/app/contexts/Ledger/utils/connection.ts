import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
