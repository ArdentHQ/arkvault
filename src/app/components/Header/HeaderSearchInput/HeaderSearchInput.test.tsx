import userEvent from "@testing-library/user-event";
import React from "react";

import { act, render, screen, waitFor } from "@/utils/testing-library";

import { HeaderSearchInput } from "./HeaderSearchInput";

describe("HeaderSearchInput", () => {
	it("should render", () => {
		const { asFragment } = render(<HeaderSearchInput />);

		expect(screen.getByTestId("HeaderSearchInput__input")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});

	it("should limit search letters", async () => {
		render(<HeaderSearchInput maxLength={32} />);

		await userEvent.click(screen.getByRole("button"));

		const input: HTMLInputElement = screen.getByTestId("HeaderSearchInput__input__input");

		expect(input.maxLength).toBe(32);

		const text = "looong text";
		const longText = text.repeat(10);

		await userEvent.paste(input, longText);

		expect(input.value).toBe(longText.slice(0, input.maxLength));
		expect(input.value).toHaveLength(input.maxLength);
	});

	it("should reset fields by prop", async () => {
		const onReset = vi.fn();
		const { rerender } = render(<HeaderSearchInput onReset={onReset} />);

		await userEvent.click(screen.getByRole("button"));

		const input: HTMLInputElement = screen.getByTestId("HeaderSearchInput__input__input");

		await userEvent.paste(input, "test");

		expect(input.value).toBe("test");

		rerender(<HeaderSearchInput onReset={onReset} resetFields={true} />);

		await waitFor(() => expect(input.value).not.toBe("test"));

		expect(onReset).toHaveBeenCalledWith();
	});

	it("should reset the query", async () => {
		const onReset = vi.fn();
		render(<HeaderSearchInput onReset={onReset} />);

		await userEvent.click(screen.getByRole("button"));

		const input: HTMLInputElement = screen.getByTestId("HeaderSearchInput__input__input");

		await userEvent.paste(input, "test");

		expect(input.value).toBe("test");

		await userEvent.click(screen.getByTestId("HeaderSearchInput__input__reset"));

		expect(input.value).not.toBe("test");
		expect(onReset).toHaveBeenCalledWith();
	});

	it("should call onSearch", async () => {
		vi.useFakeTimers();

		const onSearch = vi.fn();

		render(<HeaderSearchInput onSearch={onSearch} />);

		await userEvent.click(screen.getByRole("button"));

		await userEvent.paste(screen.getByTestId("HeaderSearchInput__input__input"), "test");

		act(() => {
			vi.runAllTimers();
		});

		expect(onSearch).toHaveBeenCalledWith("test");
	});

	it("should set custom debounce timeout form props", async () => {
		vi.useFakeTimers();

		const onSearch = vi.fn();

		render(<HeaderSearchInput onSearch={onSearch} debounceTimeout={100} />);

		await userEvent.click(screen.getByRole("button"));

		await userEvent.paste(screen.getByTestId("HeaderSearchInput__input__input"), "test");

		act(() => {
			vi.runAllTimers();
		});

		expect(onSearch).toHaveBeenCalledWith("test");
	});
});
