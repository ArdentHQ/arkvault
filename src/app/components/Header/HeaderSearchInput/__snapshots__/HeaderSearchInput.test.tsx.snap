// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderSearchInput > should render 1`] = `
<DocumentFragment>
  <div
    class="relative"
    data-testid="HeaderSearchInput"
  >
    <div
      class="dark:border dark:border-theme-secondary-800 xs:[min-width:300px] sm:[min-width:448px] flex items-center overflow-hidden rounded-lg border border-theme-secondary-400 bg-theme-background px-4 text-base dark:border-theme-secondary-700"
      data-testid="HeaderSearchInput__input"
    >
      <button
        class="transition-all duration-300 focus:outline-none"
        data-testid="HeaderSearchInput__input__reset"
        type="button"
      >
        <div
          class="text-theme-text transition-all duration-300 w-0"
        >
          <div
            style="height: 16px; width: 16px;"
          >
            <svg
              id="cross"
              style="height: 100%; width: 100%;"
              viewBox="0 0 20 20"
              x="0"
              xml:space="preserve"
              xmlns="http://www.w3.org/2000/svg"
              y="0"
            >
              <g
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              >
                <path
                  d="M.954 18.981l18-18M18.954 18.981l-18-18"
                />
              </g>
            </svg>
          </div>
        </div>
      </button>
      <div
        class="flex-1"
      >
        <div
          class="-ml-4 css-1sdc56v"
        >
          <div
            class="relative flex h-full flex-1"
          >
            <input
              autocomplete="off"
              class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
              data-testid="HeaderSearchInput__input__input"
              placeholder="Search..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="text-theme-secondary-700 dark:text-theme-secondary-600"
      >
        <div
          style="height: 20px; width: 20px;"
        >
          <svg
            fill="none"
            style="height: 100%; width: 100%;"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_13375_36442)"
              id="Vertical container"
            >
              <g
                id="Vertical container_2"
              >
                <path
                  d="M10.5038 13.9035C13.7838 12.5435 15.3038 8.70355 13.9438 5.50355C12.5838 2.22355 8.74379 0.783547 5.54379 2.06355C2.34379 3.34355 0.743788 7.26355 2.10379 10.4635C3.54379 13.7435 7.22379 15.2635 10.5038 13.9035Z"
                  id="Vector"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.6"
                />
                <path
                  d="M12.8281 12.7832L14.4281 14.3832"
                  id="Vector_2"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.6"
                />
              </g>
            </g>
            <defs>
              <clippath
                id="clip0_13375_36442"
              >
                <rect
                  fill="transparent"
                  height="16"
                  width="16"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
